### DISCUSSION TITLE
How to create a custom cursor?

### DISCUSSION THREAD

**Discussion by ricd** [microsoft/WindowsAppSDK#2722] [Q&A] [answered]
Setup:
Latest Windows App SDK (10.0.25151-preview).  A C#, Windows App SDK, Win UI 3 desktop application.

I need to change the cursor as the user interacts with a Grid.  I have sub-classed the Grid in order to provide access to its ProtectedCursor:
```
internal class Grid2 : Grid
{
	public InputCursor Cursor
	{
		get => ProtectedCursor;

		set => ProtectedCursor = value;
	}
}
```
I have created a .res file containing cursors (.cur) in a separate C++ DLL project and added that resource file as a Win32Resouce to my Windows App SDK/WinUI 3 desktop application project.

As a test, I am using the PointerPressed event from the subclassed Grid to change the cursor:
```
private void OnPointerPressed(object sender, PointerRoutedEventArgs e)
{
       // 102 is the resource id of a cursor in the .res file.
	var cursor = new CoreCursor(Windows.UI.Core.CoreCursorType.Custom, 102);

	var t = InputCursor.CreateFromCoreCursor(cursor);
	_rootGrid.Cursor = t;
}
```
However, the effect of the code above is to simply load a built-in cursor.  Indeed, as I play with the resource id value, setting it to values in the range [100, 118] I see a variety of 'standard' cursors including some that are not in the CoreCursorType enumeration.

The same .res file works in a UWP application, so I know the problem is not with the resources...  

If I change the resource id of one of the cursors to, say, 200, recreate the .res file and try to create a CoreCursor using that resource Id, I get a System.Runtime,InteropServices.COMException: The Specified resource name cannot be found in the image file.

Why  when creating a new CoreCursor  with CoreCursorType.Custom, and a uint resource Id am I seeing build-in cursors?

**Comment by castorix** [comment]
I cannot even use _Win32Resource_ in the .csproj to add a .res on my configuration (Windows 10 21H1, unpackaged)
(error 0xc0000409 on running)
I also tried with a DLL with _Microsoft.UI.Input.InputDesktopResourceCursor.CreateFromModule_,
but it does not find the DLL and then calls [LoadCursor](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-loadcursorw) with NULL, so it loads a System cursor

**Comment by Kevin-Taha** [comment]
Some resource IDs like 100, 101... are reserved by system resources (in the current executable module), which is why you see it loading syscursors at those IDs and failing with less generic IDs. When you call `CreateFromCoreCursor` with a Custom CoreCursor, the method forwards on to the constructor for InputDesktopResourceCursor (with a null module name b/c CoreCursor only supports the current module unlike InputCursor), which succeeds or fails dependent on the output of [LoadCursor](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-loadcursora).

If you want to boil it down a little more, I'd suggest trying to load a resource from that linked DLL directly using LoadCursor, LoadResource, or a similar Win32 Resource API to deduce if the resources are properly linked to the project. Imo while resource linking works great in a C++ Proj, the flow of linking Win32 resources in a C# project is not particularly intuitive or consistent, and I've had similar frustrations 😞 

You may be able to directly create an InputCursor from your resource by calling InputDesktopResourceCursor.CreateFromModule and passing the name of the DLL as the module. If the DLL is loaded, LoadCursor should be able to get the right resource from it.  CreateFromCoreCursor is simply a convenience method that does nothing which InputCursor can't already do.

**Reply by ricd** [reply]
Kevin, thanks for the reply. Your info along with the reply from Marv51 provided the info I needed.

**Comment by Marv51** [comment, marked as answer]
I had the same issue, my cursors refused to load, but I fixed it like this:

First, To get the .res file in my c# dll, I specify `<Win32Resource>MyResources.res</Win32Resource>`.
And I removed the `<Manifest Include="$(ApplicationManifest)" />` and `<ApplicationManifest>app.manifest</ApplicationManifest>` properties.
To add the manifest back, I added the manifest to my MyResources.res file, via "Import", and set the type to "RT_MANIFEST".
On the res file I left the the Build Action as "None" and "Copy to Output Directory" as "Do not copy".


Second, I have validated with [ResourceHacker](http://www.angusj.com/resourcehacker/) that a cursor with the expected id exists in the DLL. I set the cursor ID in the DLL outside of the 1xx range to avoid the system cursors.


Third, instead of using CoreCursor, I call `InputDesktopResourceCursor.CreateFromModule` with my app dll specified as the module name:
```csharp
var cursor = InputDesktopResourceCursor.CreateFromModule("App1.dll", 217);
```

That is all that it took for me, the cursor now shows as expected.

**Reply by ricd** [reply]
Marvin,  Thank you.  Your reply along with the info from Kevin helped resolve the problem.

**Reply by benstevens48** [reply]
Thanks for this. One additional thing I noticed is that I had to set the ID of the RT_MANIFEST resource to 1 otherwise it was ignored (DPI was not correct), or at least with an ID of 101 it was ignored (I didn't try other IDs).

**Reply by benstevens48** [reply]
Actually, one more additional thing, it turns out if you are publishing with Native AOT then you need to use "App1.exe" instead of "App1.dll" for the published app, but you need to add a fallback to "App1.dll" for it to work in debug mode as well. (In published Native AOT apps there is no "App1.dll", it gets combined into "App1.exe").

**Comment by KusiTec** [comment]
Thanks a lot for the solution

**Comment by BlameTwo** [comment]
他是否支持动态光标？

**Reply by castorix** [reply]
> 他是否支持动态光标？

Not sure if it is what you mean, but you can create dynamically a Win32 cursor, then call [IInputCursorStaticsInterop::CreateFromHCursor method](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/win32/microsoft.ui.input.inputcursor.interop/nf-microsoft-ui-input-inputcursor-interop-iinputcursorstaticsinterop-createfromhcursor)
